File: DoAnChuyenNganh-main/DoAnCoSo/WebDuLichDaLat/Models/TouristPlace.cs
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;

namespace WebDuLichDaLat.Models
{
    public class TouristPlace
    {
        [Required(ErrorMessage = "Mã địa điểm không được để trống")]
        [StringLength(6, MinimumLength = 6, ErrorMessage = "Phải có đúng 6 kí tự.")]
        public required string Id { get; set; }

        [Required(ErrorMessage = "Tên địa điểm không được để trống")]
        [StringLength(50, ErrorMessage = "Tên địa điểm không được quá 50 kí tự")]
        public required string Name { get; set; }

        public int? RegionId { get; set; }
        public virtual Region? Region { get; set; }

        public string? ImageUrl { get; set; }

        public List<string>? ImageUrls { get; set; }

        public string? Description { get; set; }
        public virtual Category? Category { get; set; }
        public int? CategoryId { get; set; }

        public double Latitude { get; set; }
        public double Longitude { get; set; }
        public string? ReviewContent { get; set; }
        public int Rating { get; set; } // Từ 1 đến 5
        public ICollection<Review> Reviews { get; set; } = new List<Review>();
        public ICollection<Hotel> Hotels { get; set; } = new List<Hotel>();
        public ICollection<Restaurant> Restaurants { get; set; } = new List<Restaurant>();
        public ICollection<TransportOption> TransportOptions { get; set; } = new List<TransportOption>();


    }
}

File: src/main/java/com/example/dalattravel/model/TouristPlace.java
package com.example.dalattravel.model;

import jakarta.persistence.*;
import lombok.*;
import java.util.ArrayList;
import java.util.List;

@Entity
@Table(name = "tourist_places")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class TouristPlace {

    @Id
    @Column(length = 6, nullable = false)
    private String id;

    @Column(length = 100, nullable = false)
    private String name;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "region_id")
    private Region region;

    @Column(name = "region_id", insertable = false, updatable = false)
    private Integer regionId;

    private String imageUrl;

    @ElementCollection
    @CollectionTable(name = "tourist_place_images", joinColumns = @JoinColumn(name = "tourist_place_id"))
    @Column(name = "image_url")
    @Builder.Default
    private List<String> imageUrls = new ArrayList<>();

    @Column(columnDefinition = "TEXT")
    private String description;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "category_id")
    private Category category;

    @Column(name = "category_id", insertable = false, updatable = false)
    private Integer categoryId;

    private double latitude;
    private double longitude;

    @Column(columnDefinition = "TEXT")
    private String reviewContent;

    private int rating;

    private java.math.BigDecimal ticketPrice;
    private Integer avgVisitDurationMin;

    @OneToMany(mappedBy = "touristPlace", cascade = CascadeType.ALL, fetch = FetchType.LAZY)
    @Builder.Default
    private List<Review> reviews = new ArrayList<>();

    @OneToMany(mappedBy = "touristPlace", cascade = CascadeType.ALL, fetch = FetchType.LAZY)
    @Builder.Default
    private List<Hotel> hotels = new ArrayList<>();

    @OneToMany(mappedBy = "touristPlace", cascade = CascadeType.ALL, fetch = FetchType.LAZY)
    @Builder.Default
    private List<Restaurant> restaurants = new ArrayList<>();
}
